File: modules/app/src/main/java/com/prevengos/plug/app/usecase/GuardarCuestionarioUseCase.java
package com.prevengos.plug.app.usecase;

import com.prevengos.plug.app.dto.GuardarCuestionarioCommand;
import com.prevengos.plug.domain.events.CuestionarioCompletadoEvent;
import com.prevengos.plug.domain.events.DomainEventPublisher;
import com.prevengos.plug.domain.model.Cuestionario;
import com.prevengos.plug.domain.ports.CuestionariosPort;

import java.util.Objects;

public class GuardarCuestionarioUseCase {
    private final CuestionariosPort cuestionariosPort;
    private final DomainEventPublisher eventPublisher;

    public GuardarCuestionarioUseCase(CuestionariosPort cuestionariosPort, DomainEventPublisher eventPublisher) {
        this.cuestionariosPort = Objects.requireNonNull(cuestionariosPort, "cuestionariosPort");
        this.eventPublisher = Objects.requireNonNull(eventPublisher, "eventPublisher");
    }

    public Cuestionario execute(GuardarCuestionarioCommand command) {
        Cuestionario cuestionario = cuestionariosPort.guardar(command.toCuestionario());
        if (cuestionario.completadoEn() != null) {
            eventPublisher.publish(new CuestionarioCompletadoEvent(command.fuente(), cuestionario));
        }
        return cuestionario;
    }
}

File: modules/app/src/main/kotlin/com/prevengos/plug/app/usecase/GuardarCuestionarioUseCase.kt
package com.prevengos.plug.app.usecase

import com.prevengos.plug.app.dto.GuardarCuestionarioCommand
import com.prevengos.plug.domain.events.CuestionarioCompletadoEvent
import com.prevengos.plug.domain.events.DomainEventPublisher
import com.prevengos.plug.domain.model.Cuestionario
import com.prevengos.plug.domain.ports.CuestionariosPort

class GuardarCuestionarioUseCase(
    private val cuestionariosPort: CuestionariosPort,
    private val eventPublisher: DomainEventPublisher
) {
    fun execute(command: GuardarCuestionarioCommand): Cuestionario {
        val cuestionario = cuestionariosPort.guardar(command.toCuestionario())
        if (cuestionario.completadoEn != null) {
            eventPublisher.publish(
                CuestionarioCompletadoEvent(
                    source = command.fuente,
                    cuestionario = cuestionario
                )
            )
        }
        return cuestionario
    }
}
